### DISCUSSION TITLE
Add sitemap GraphQL query

### DISCUSSION THREAD

**Discussion by el-schneider** [aerni/statamic-advanced-seo#88] [Ideas]
It would be nice if a flat XML sitemap could be queried through GraphQL

**Comment by aerni** [comment]
I'm not sure how the sitemap based on your Statamic routes is relevant in a headless setup, as routing is taken care of by the frontend framework. How is Advanced SEO supposed to know about your frontend routing? In React or Vue land you would use a package that generates the sitemap based on your frontend routes.

Also, what do you mean by `flat XML sitemap`. Do you have an example?

**Comment by el-schneider** [comment]
I think more often than not frontend routing somehow relies on the data structure in the backend. In my case I have a nested collection `pages`, that defines its routes like so: `{parent_uri}/{slug}`. In my frontend app, SvelteKit in my case, but this should apply to Nuxt, Next and so on as well, I would query such an entry base on a catchall url wildcard, that I pass as an argument to my GraphQL query:
```graphql
query Page($uri: String!) {
  entry(uri: $uri) {
    id
  }
}
```

I could of course make a couple of queries and merge date I get from  `seoDefaults` and `entries > data > seo > computed > indexing` to figure out which entries should be included in the sitemap, depending on the settings in the CP, but all this logic has already been implemented by you at some point, so I much rather would like to query all the entries, or a prerended XML sitemap through one single query. On that note: I think a top-level field `seoSitemap`, that implements three fields, `computed`, `raw`and `view` following the implementation of your `seo` field would probably work best and cover all the use-cases.

I'm no expert in React or Vue land, but here is a a link to the NuxtSeo Docs, talking about generating a Sitemap for dynamic URLs https://nuxtseo.com/sitemap/guides/dynamic-urls, which basically involves returning an array of locations with fields `loc`, `lastmod`, `changefreq` and `priority`


With **flat** XML sitemap I'm referring to just one XML document `sitemap.xml` containing all the urls, as opposed to what `Advanced SEO` is currently providing: an index `sitemap.xml`, that links to sub-sitemaps like `sitemap-collection-pages.xml`

**Comment by aerni** [comment]
For the sitemap to be useful, your frontend routes have to be perfectly synced with Statamic's routes. Is that realistic? Imagine you've got one additional route that isn't taken care of by Statamic. Now you've got an incomplete sitemap. Of course, that's only the case with a rendered XML. If you could query the sitemap data, you could still add and remove items and take care of the XML yourself. I think this could make sense.

As for the flat sitemap, why would you need this?

**Comment by el-schneider** [comment]
Having frontend-routes in sync with backend routes is definitely the case for me. I also have a strong incentive here, as by using your addon I'm enabling my client to control seo meta data on every page, through some collection entry. Admittedly mine might be a special use-case and I would be also very glad if instead of an sitemap `view` I could just get a top-level field `seoSitemap` with `loc`, `lastmod`, `changefreq` and `priority`, that adheres to whatever has been set in `/cp/advanced-seo/site/indexing` or been overwritten by an individual entry.

**Comment by aerni** [comment]
I'm adding this to my list of feature requests. 

I won't implement the "flat" sitemap, though. Separating sitemaps was an architectural decision to improve performance. And nested sitemaps work fine. I don't see a reason why you'd need a flat sitemap.

**Comment by el-schneider** [comment]
@aerni Great, thank you! much appreciated ...

**Comment by maartenvanhunsel** [comment]
When we create (headless) websites, Statamic is our source of truth and collections urls are configured from Statamic. Custom (FE-only) pages can be added manually by the frontend module.

For example we use Nuxt and Nuxt simple sitemap which supports dynamic URL's. (https://nuxtseo.com/sitemap/guides/dynamic-urls)
The only thing we need from the API is the sitemap structure in JSON filled with `slug` or `url`, `updated_at`, `changefreq` and `priority` (see https://nuxtseo.com/sitemap/guides/dynamic-urls#build-time-dynamic-urls).

For us it doesn't matter if the sitemap is 'flat' or not, important is that we can do whatever what we want with it. 
Also I'm not a SEO expert but the nuxtseo package recommends not chunking sitemaps [unless more then 1000 urls](https://nuxtseo.com/sitemap/guides/multi-sitemaps#automatic-chunking).

**Comment by aerni** [comment]
I've got a first implementation going. How does the following look to you? Any additional arguments or filters that would make sense on the query?

```graphql
query MyQuery {
  seoSitemap {
    alternates {
      href
      hreflang
    }
    changefreq
    lastmod
    loc
    priority
  }
}
```

```json
{
  "data": {
    "seoSitemap": [
      {
        "alternates": [
          {
            "href": "https://statamic-advanced-seo.test/about",
            "hreflang": "en-US"
          },
          {
            "href": "https://statamic-advanced-seo.test/german/about",
            "hreflang": "de-DE"
          }
        ],
        "changefreq": "daily",
        "lastmod": "2023-05-15T20:05:17+00:00",
        "loc": "https://statamic-advanced-seo.test/german/about",
        "priority": "0.5"
      },
      {
        "alternates": [
          {
            "href": "https://statamic-advanced-seo.test/about",
            "hreflang": "en-US"
          },
          {
            "href": "https://statamic-advanced-seo.test/german/about",
            "hreflang": "de-DE"
          }
        ],
        "changefreq": "daily",
        "lastmod": "2023-05-15T20:19:38+00:00",
        "loc": "https://statamic-advanced-seo.test/about",
        "priority": "0.5"
      },
      {
        "alternates": [],
        "changefreq": "daily",
        "lastmod": "2023-05-16T13:05:10+00:00",
        "loc": "https://statamic-advanced-seo.test",
        "priority": "0.5"
      },
      {
        "alternates": [],
        "changefreq": "daily",
        "lastmod": "2023-07-28T13:46:09+00:00",
        "loc": "https://statamic-advanced-seo.test/tags",
        "priority": "0.5"
      }
    ]
  }
}
```

**Comment by el-schneider** [comment]
Looking good to me!

**Comment by aerni** [comment]
Wouldn't you need a `baseUrl` argument so you can change the URL of the `loc` field?

**Reply by maartenvanhunsel** [reply]
@aerni Yes that's true, you have to point out to the frontend URL.

**Comment by aerni** [comment]
Any other query arguments that could be useful? Like only querying sitemap items of a certain collection. Or filters to only get items with a certain priority? Just brainstorming here before finalizing the implementation.

**Reply by maartenvanhunsel** [reply]
Filtering per collection would be great, without the filter::collection you just get all collections right?
What about multi-site filtering?

**Reply by aerni** [reply]
Right now, you get all the sitemap URLs, including entries, terms, and collection terms. If an entry/term is localized in another site you'll get an array of `alternates`.

It probably makes sense to add filters for `collection`, `taxonomy`, and `site`. Any other thing that comes to mind?

**Reply by maartenvanhunsel** [reply]
That's it for now.

**Comment by aerni** [comment]
Could you give PR #96 a try before I tag a release?
